### ISSUE TITLE
Bug in .NET with a number comparison

### ISSUE THREAD

**Issue by chschrae** [jdevillard/JmesPath.Net#52]
Hi there,

I believe there is a bug in the transform of the .NET code when using an expression with an && and a number comparison. Consider the following code
```
            string jsonString = @"{
                ""StringKey"": ""StringValue"",
                ""StringKey2"": ""StringValue2"",
                ""NumberKey"": 100}";
            string jmesExpression = "StringKey == 'StringValue' && NumberKey > '5'";
            var jmes = new JmesPath();

            var jmesResult = jmes.Transform(jsonString, jmesExpression);

            Console.WriteLine($"JMES result string = {jmesResult}");

            string jmesExpression2 = "StringKey == 'StringValue' && StringKey2 == 'StringValue2'";

            var jmesResult2 = jmes.Transform(jsonString, jmesExpression2);

            Console.WriteLine($"JMES result 2 string = {jmesResult2}");

            string jmesExpression3 = "NumberKey > '5'";

            var jmesResult3 = jmes.Transform(jsonString, jmesExpression3);

            Console.WriteLine($"JMES result 3 string = {jmesResult3}");

            string jmesExpression4 = "NumberKey == '100'";

            var jmesResult4 = jmes.Transform(jsonString, jmesExpression4);

            Console.WriteLine($"JMES result 4 string = {jmesResult4}");
```

If you run this, you will see that the first transform where the NumberKey is compared will return null and the second transform will return true. The third also returns null and the fourth returns false so there is something wrong with the number comparison.

**Comment by chschrae** [comment]
I realized that I should be using backticks instead of single quotes, but there is something weird on the example web page where single quotes work sometimes
![image](https://user-images.githubusercontent.com/32145632/173656820-ef71f901-8566-4b3e-b11d-eaea1fb00048.png)

**Comment by springcomp** [comment]
Hi @chschrae Thanks for your feedback.

You are right, there is a bug, but not where you think there is 😉.

In fact, the specification clearly states that equality operators should apply an exact match. For this reason, comparing `'100'` with `` `100` `` is actually comparing a string to a number and that returns `false` as per the specification.

On the other hand, the specification states that ordering operators are only valid for numbers and that evaluating any other type with a comparison operator should yield a `null` value. You can cleary see that it returns a "correct" result of comparing numbers in your example when applying the expression `NumberKey > '5'`. This is actually a deviation from the standard. However, the .NET implementation tries to be as faithful as possible to the JavaScript implementation that powers examples from the website. There, the comparison succeeds, so we decided to apply a conversion to numbers and make the comparison if that still made sense to mimic this behaviour.

In short:

`search(NumberKey == '100', {"NumberKey": 100}) -> false` is compliant with the specification.
`search(NumberKey > '5', {"NumberKey": 100}) -> true` is not, but returns the intuitively correct answer 😉

**Comment by chschrae** [comment]
Oh wow! You are absolutely right! I didn't even realize the back ticks were a separate operator. Thanks for responding!
